actions/hdl_example/sw/snap_example_qnvme.c

```diff
@@ -0,0 +1,659 @@
+#define    CMD_ACTION_ID(aid)     (aid << 8)
+
+/* NVME Read Registers */
+#define STATUS_REG   0x00
+#define   SUB_Q_FULL_SSD0_A       0x01
+#define   SUB_Q_FULL_SSD0_IO      0x02
+#define   SUB_Q_FULL_SSD1_A       0x04
+#define   SUB_Q_FULL_SSD1_IO      0x08
+#define TRACK_REG    0x04
+
+static int nvme_qtest(struct snap_card *handle, /* My handle */
+		bool write,                     /* Set to true if doing write */
+		uint64_t ram_address,           /* Set any ram address for sorce ore dest buffer */
+		int drive,                      /* Use SSD drive 0 or 1 */
+		int blocks,                     /* How many blocks to write or read */
+		int blocks_offset,              /* Offset (block#) to start write or read */
+		int nmax,                       /* queue counter */
+		int use_aid)                    /* Action id */
+{
+	int qin, n;
+	uint32_t dptr_low;
+	uint32_t dptr_high;
+	uint32_t lba_low = blocks_offset;           /* Lba to Start */
+	uint32_t ssd_ioq_full = SUB_Q_FULL_SSD0_IO; /* Mask to check for drive 0 */
+	uint32_t cmd_reg = 0;                       /* Command Reg */
+	uint32_t val32;                             /* Scratch data */
+	uint32_t aid_val = 0x0;                     /* Action Value */
+	uint32_t track_reg = TRACK_REG * (use_aid + 1);
+	uint64_t t_start = 0;                           /* time in usec */
+	uint64_t t_end;                             /* End in usec */
+	uint64_t t_q;                             /* End in usec */
+	uint64_t t_0;                               /* q in usec */
+	uint64_t t_1;                               /* q in usec */
+	uint64_t total_size = (uint64_t)(nmax * blocks * NVME_LB_SIZE);
+	int t_i = 0;
+
+	int *t_dlist = malloc(nmax * sizeof(int));
+	VERBOSE2("\n%s Enter %s SSD[%d] Blocks: %d (%d Bytes) N: %d (Total: %lld Bytes) AID: %d\n"
+		"   Using RAM: 0x%016llx\n",
+		__func__, write?"Write":"Read", drive, blocks,
+		blocks * NVME_LB_SIZE, nmax,
+		(long long)total_size, use_aid,
+		(long long)ram_address);
+
+	cmd_reg = CMD_QUEUE_ID_SSD0_IOQ | CMD_TYPE_READ;
+	if (1 == drive) {
+		ssd_ioq_full = SUB_Q_FULL_SSD1_IO;
+		cmd_reg = CMD_QUEUE_ID_SSD1_IOQ;
+	}
+	if (write)
+		cmd_reg |= CMD_TYPE_WRITE;
+
+	cmd_reg |= CMD_ACTION_ID(use_aid);
+
+	/* Make values for DPTR_HIGH Note the offset is 8 GB, so i add 2 */
+	dptr_high = 2 + (uint32_t)(ram_address >> 32ll);
+	/* Make values for DPTR_LOW */
+	dptr_low = (uint32_t)ram_address;
+
+	VERBOSE3("   Write DPTR_HIGH: 0x%x\n", dptr_high);
+	nvme_mmio_write(handle, DPTR_HIGH, dptr_high); /* DPTR_HIGH Transfer data pointer high 32 bits */
+	VERBOSE3("   Write LBA_HIGH:  0\n");
+	nvme_mmio_write(handle, LBA_HIGH, 0);          /* LBA_High SSD LBA high 32 bit */
+	VERBOSE3("   Write LBA_NUM:   %d\n", blocks-1);
+	nvme_mmio_write(handle, LBA_NUM, blocks-1);    /* LBA_NUM Number of LBA Block in transfer 0..65536 */
+	VERBOSE3("   Using Track Reg: 0x%x\n", track_reg);
+	qin = 0;
+	t_0 = get_usec();
+	t_start = t_0;
+	for (n = 0; n < nmax; n++) {
```

Comment:
I didn't execute it, but the loop seems to send only (nmax-1) requests.

Reply:
You have to set --nblk or -n option to set nmax.